crypto/rand/rand_unix.c

```diff
@@ -325,6 +325,44 @@ static struct random_device {
 } random_devices[OSSL_NELEM(random_device_paths)];
 static int keep_random_devices_open = 1;
 
+/*
+ * Ensure that the system randomness source has been adequately seeded.
+ * This is done by having the first start of libcrypto, read one bytes from
+ * /dev/random.  Subsequent starts of the library and later reseedings do not
+ * need to do this.
+ */
+static void wait_random_seeded(void)
+{
+    static const char random_path[] = "/dev/random";
+    static const char seeded_file[] = "/tmp/.openssl_seeded";
+    static int seeded = 0;
+    struct stat st;
+    int fd = -1, rfd = -1;
+    char c;
+
+    if (!seeded && (stat(seeded_file, &st) == -1 || st.st_size == 0)) {
+        fd = open(seeded_file, O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR);
+        if (fd == -1 || flock(fd, LOCK_EX) == -1 || fstat(fd, &st) == -1)
+            goto err;
+        if (st.st_size == 0) {
+            rfd = open(random_path, O_RDONLY);
+            if (rfd == -1 || read(rfd, &c, 1) != 1 || write(fd, &c, 1) != 1)
+                goto err;
+            close(rfd);
+            close(fd);
+            seeded = 1;
+        }
+    }
+    return;
```

Comment:
I wonder, if two processes using libcrypto are started at the same time, and one of them creates that file and waits for /dev/random to respond, what happens with the other process?  Does it wait for the first process to be done with this part, or does it simply fail?

Reply:
The `flock(fd, LOCK_EX)` takes care of this, other processes wait.

Reply:
Ah, okie.